test/components/componentTests.ts

```diff
@@ -77,114 +90,209 @@ describe("Component behavior", () => {
 
-    it("components can be detach()-ed even if not anchor()-ed", () => {
-      let c = new Plottable.Component();
+    it("does not error if not already anchored", () => {
+      assert.doesNotThrow(() => c.detach(), Error, "does not throw error if detach() called before anchor()");
+      c.destroy();
+      svg.remove();
+    });
+
+    it("removes the parent when detaching", () => {
+      let parent = new Plottable.Components.Group([c]);
+      c.detach();
+      assert.isNull(c.parent(), "parent removed upon detaching");
+      parent.destroy();
+      c.destroy();
+      svg.remove();
+    });
+
+    describe("detach hooks", () => {
+      it("can undergo set behavior upon detaching", () => {
+        c.renderTo(svg);
+
+        let callbackCalled = false;
+        let passedComponent: Plottable.Component;
+        let callback = (component: Plottable.Component) => {
+          callbackCalled = true;
+          passedComponent = component;
+        };
+        assert.strictEqual(c.onDetach(callback), c, "setter returns calling object");
+        c.detach();
+        assert.isTrue(callbackCalled, "callback was called when the Component was detached");
+        assert.strictEqual(passedComponent, c, "callback was passed the Component that detached");
+        c.destroy();
+        svg.remove();
+      });
 
-      let callbackCalled = false;
-      let passedComponent: Plottable.Component;
-      let callback = (component: Plottable.Component) => {
-        callbackCalled = true;
-        passedComponent = component;
+      it("calls callbacks upon detaching even if not anchored", () => {
+        let callbackCalled = false;
+        let passedComponent: Plottable.Component;
+        let callback = (component: Plottable.Component) => {
+          callbackCalled = true;
+          passedComponent = component;
+        };
+        c.onDetach(callback);
+
+        c.detach();
+        assert.isTrue(callbackCalled, "callback still called");
+        assert.strictEqual(passedComponent, c, "callback passed the Component that detached");
+        c.destroy();
+        svg.remove();
+      });
+
+      it("can remove callbacks that would have been called upon detaching", () => {
+        let callbackCalled = false;
+        let callback = (component: Plottable.Component) => {
+          callbackCalled = true;
+        };
+        c.onDetach(callback);
+        assert.strictEqual(c.offDetach(callback), c, "setter calls calling object");
+        c.renderTo(svg);
+        c.detach();
+        assert.isFalse(callbackCalled, "removed callback is not called");
+        c.destroy();
+        svg.remove();
+      });
+    });
+  });
+
+  describe("parent container", () => {
+    it("can set its parent to a container that contains this component", () => {
+      let acceptingContainer = {
+        has: () => true
       };
-      c.onDetach(callback);
+      assert.strictEqual(c.parent(<any> acceptingContainer), c, "setter returns calling object");
```

Comment:
is it possible to declare acceptingContainer as any rather than casting it here?